src/ipc/ipc-zephyr.c

```diff
@@ -59,33 +63,105 @@ LOG_MODULE_DECLARE(ipc, CONFIG_SOF_LOG_LEVEL);
  */
 static uint32_t g_last_data, g_last_ext_data;
 
+struct k_sem *wait_ack_sem;
+
 /**
- * @brief cAVS IPC Message Handler Callback function.
+ * @brief cAVS IPC Message Received Callback function.
  *
- * See @ref (*intel_adsp_ipc_handler_t) for function signature description.
- * @return false so BUSY on the other side will not be cleared immediately but
+ * @return -1 so BUSY on the other side will not be cleared immediately but
  * will remain set until message would have been processed by scheduled task, i.e.
  * until ipc_platform_complete_cmd() call.
  */
-static bool message_handler(const struct device *dev, void *arg, uint32_t data, uint32_t ext_data)
+static void ipc_receive_cb(const void *data, size_t len, void *priv)
 {
-	struct ipc *ipc = (struct ipc *)arg;
+	struct intel_adsp_ipc_ept_priv_data *priv_data =
+		(struct intel_adsp_ipc_ept_priv_data *)priv;
 
-	k_spinlock_key_t key;
+	if (len == INTEL_ADSP_IPC_CB_MSG) {
+		const struct intel_adsp_ipc_msg *msg = (const struct intel_adsp_ipc_msg *)data;
 
-	key = k_spin_lock(&ipc->lock);
+		struct ipc *ipc = (struct ipc *)priv_data->priv;
 
-	g_last_data = data;
-	g_last_ext_data = ext_data;
+		k_spinlock_key_t key;
+
+		key = k_spin_lock(&ipc->lock);
+
+		g_last_data = msg->data;
+		g_last_ext_data = msg->ext_data;
 
 #if CONFIG_DEBUG_IPC_COUNTERS
-	increment_ipc_received_counter();
+		increment_ipc_received_counter();
 #endif
-	ipc_schedule_process(ipc);
+		ipc_schedule_process(ipc);
 
-	k_spin_unlock(&ipc->lock, key);
+		k_spin_unlock(&ipc->lock, key);
 
-	return false;
+		priv_data->cb_ret = -1;
+	} else if (len == INTEL_ADSP_IPC_CB_DONE) {
+		if (wait_ack_sem)
+			k_sem_give(wait_ack_sem);
+
+		priv_data->cb_ret = INTEL_ADSP_IPC_CB_RET_OKAY;
+	}
+}
+
+static struct ipc_ept ipc_ept;
+static struct intel_adsp_ipc_ept_priv_data ipc_ept_priv_data;
+static struct ipc_ept_cfg ipc_ept_cfg = {
+	.name = "sof_ipc",
+	.cb = {
+		.received = ipc_receive_cb,
+	},
+	.priv = (void *)&ipc_ept_priv_data,
+};
+
+static void ipc_ept_init(struct ipc *ipc)
+{
+	int ret;
+	const struct device *ipc_dev = INTEL_ADSP_IPC_HOST_DEV;
+
+	ipc_ept_priv_data.priv = (void *)ipc;
+
+	ret = ipc_service_register_endpoint(ipc_dev, &ipc_ept, &ipc_ept_cfg);
+	ARG_UNUSED(ret);
+}
+
+static void ipc_complete(void)
+{
+	int ret;
+
+	ret = ipc_service_send(&ipc_ept, NULL, INTEL_ADSP_IPC_SEND_DONE);
+
+	ARG_UNUSED(ret);
```

Comment:
as long as `ipc_service_send()` isn't marked for compulsory return value check, it's ok to just call it without assigning its return value to a variable